### PR TITLE
[SPARK-21248][SS]The clean up codes in StreamExecution should not be interrupted

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

This PR uses `runUninterruptibly` to avoid that the clean up codes in StreamExecution is interrupted. It also removes an optimization in `runUninterruptibly` to make sure this method never throw `InterruptedException`.

## How was this patch tested?

Jenkins

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/util/UninterruptibleThreadSuite.scala [hunk 1]
```diff
@@ -80,8 +79,8 @@ class UninterruptibleThreadSuite extends SparkFunSuite {
     t.interrupt()
     interruptLatch.countDown()
     t.join()
-    assert(hasInterruptedException === true)
```
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/util/UninterruptibleThreadSuite.scala]: Updated this test to match the new behavior.

**Comment by zsxwing** [comment]
cc @tdas

File: core/src/test/scala/org/apache/spark/util/UninterruptibleThreadSuite.scala [hunk 2]
```diff
@@ -68,7 +68,6 @@ class UninterruptibleThreadSuite extends SparkFunSuite {
         Uninterruptibles.awaitUninterruptibly(interruptLatch, 10, TimeUnit.SECONDS)
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/util/UninterruptibleThreadSuite.scala]: does it make sense to even have this test any more?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 3]
```diff
@@ -357,7 +357,7 @@ class StreamExecution(
         if (!NonFatal(e)) {
           throw e
         }
-    } finally {
+    } finally microBatchThread.runUninterruptibly {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: can you add docs regarding why this is important?

File: core/src/test/scala/org/apache/spark/util/UninterruptibleThreadSuite.scala (hunk 2 above)
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/util/UninterruptibleThreadSuite.scala]: Yes. This test is testing if we keep the interrupted state properly now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 3 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: done

**Comment by tdas** [comment]
LGTM.

**Comment by zsxwing** [comment]
Thanks! Merging to master.

**Pull request closed without merging**